scripts/validate_rst_title_capitalization.py

```diff
@@ -121,6 +175,7 @@ def find_titles(rst_file: str) -> Generator[Tuple[str, int], None, None]:
                 len(line_chars) == 1
                 and line_chars.pop() in symbols
                 and len(line) == len(previous_line)
+                and previous_line[0] != ":"
```

Comment:
Sorry, I think I suggested this approach myself, but I see a problem now. This will skip the whole line from validation, and I see we use a mix of these `:class:` directives and regular text in some titles.

Instead of skipping the line here, I think we should update the condition in line 139. So, we check that `word` is not one of the exceptions, and also, it doesn't start with `:`. We could use a regex probably, but let's start simple, and see if it's enough.

What do you think?

Reply:
The problem is that:

- First of all this part of code line 180

    > yield re.sub(r"[`\*_]", "", previous_line), i

    removes "`" from the title. So when function checks if title is correct, it will already be modified. So this is the main problem.

- Secondly, line 129:

    `correct_title: str = re.sub(r"^\W*", "", title).capitalize()`

    removes all non word characters at the beginning from the desired correct title that we want to keep, so ":" is not kept as being part of the correct title

- Thirdly, this line, number 136

    `word_list = re.split(r"\W", removed_https_title)`

    removes all non word characters from the list to analyse. So ":" don't appear anymore.

By the way, code in line 180 causes problems because it also removes "_" which is present in documentation like > code_style.rst line 21

`Using foo.__class__`

So what I suggest is to modify in script "validate_rst_title_capitalization.py" from line 180 on:

NB: I replace "`" by {backtick} as it causes trouble to display it as code

`yield re.sub(r"[{backtick}\*_]", "", previous_line), i`

to:

`if previous_line[0] != ":": yield re.sub(r"[{backtick}\*_]", "", previous_line), i
else: yield re.sub(r"[\*_]", "", previous_line), i`

in order to keep the "`" if title begins by a ":"
And to add in "validate_rst_title_capitalization.py" line 126:

`if title[0] == ":": return title`

So that if title begins by ":" it considers title to be valid no matter what.

Or to simply add in "validate_rst_title_capitalization.py" line 126:

`if title[0] == ":": return title`

for the same reason.


Note that there is still a problem with:

> yield re.sub(r"[`\*_]", "", previous_line), i

That doesn't keep "_" in the titles